text/0000-debuginfo-based-panic.md

```diff
@@ -0,0 +1,115 @@
+* Cargo's release profile enables external debug info.
+* Panic messages now always include a backtrace (instead of the old location string).
+	* This means that the root cause of panics from `unwrap()` and similar interfaces are now much clearer.
+	* It's simple and obvious as most modern languages print backtraces for unhandled errors.
+	* The backtrace is generated from debug information. Without debug symbols, memory addresses are printed instead. The standard `addr2line` utility can be used to manually obtain location information once debug info is available.
+	* User perspectives:
+		* You're developing an application. You build in the `dev` profile. You have debug symbols and thus complete backtraces for every panic.
+		* You're testing your application (think QA). It's a `release` build. You have debug symbols and thus complete backtraces for every panic.
+		* You ship your application to an end user. You ship only the binary, not the symbol file. The end user runs into a panic and receives a raw stack trace (of memory addresses). The end user now reports the bug to you. Using `addr2line` you obtain a complete backtrace.
+		* Alternatively, you ship the symbol file to the user. The user can now debug your application without recompiling and - obviously - gets complete backtraces for every panic.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+**Symbol files** already exist (and work) on Windows and MacOS. On Linux, it's as simple as writing the main binary to one ELF image and all the debug sections to another. The existing `-C debuginfo` compiler options gains a third level "3 = full debug info in an external symbol file". The cargo release profile sets `debuginfo=3` (instead of `debuginfo=0`).
+
+The default panic handler no longer uses the location information supplied by `panic!()`. Instead, it always displays a backtrace. If there is no debug info in the binary itself, it looks for symbol files and loads those (this already works on Windows and on MacOS since [rust-lang/rust#44251](https://github.com/rust-lang/rust/pull/44251), still missing for Linux though).
+
+It does so through an interface in `std::panic` by first obtaining a raw stack trace (basically `Iterator<usize>`) and then requesting symbols for those stack frames. This interface can be very useful for projects like error-chain as well, especially since *obtaining* the raw stack trace is very *cheap*.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The single major drawback here is implementation complexity. Walking the stack is usually quite simple but parsing debug info to obtain symbols is not. On the other hand, backtraces already work on most platforms and are even essential to debugging for many users, so most of this is either already solved or worth the effort on its own.
```

Comment:
Debuginfo is commonly 3-5x the size of the main binary. Relying on that much stuff to get line information seems like a drawback to me.

For reference, we build a service at work with level 1 debuginfo to have symbolicated stack traces. The binary is 53 MiB, of which 41 MiB is debuginfo.

Reply:
On the other hand debuginfo can easily be left out, whereas line information hardcoded into the binary is always there no matter what you do wasting space.

Reply:
How much space do those hardcoded lines take up? Has anyone proposed a flag to turn off `line()` &co?

Reply:
This RFC proposes to **replace** the `file!()`+`line!()` panic locations with backtraces so it's always "turned off" for panics.

I'm aware that debuginfo can be **huge**, I just don't see why that's a problem:

* This RFC doesn't change anything for debug builds.
* In release builds:
    * Compile time is a lot less important
    * Size of the binary itself doesn't change (debug info is external)
    * Sure, the symbol file consumes disk space. But when is that a problem? You don't usually ship it to customers/production.
        * In cases where you do want debuginfo in production for some reason (like the example you mentioned), you can still just ask the compiler to give you what you want.

So what does change?

 * By default, you no longer print panic locations in production / to customers. The underlying rationale of both this RFC and #2091 (that, I think, everyone agrees with at this point) is that today's panic locations are rarely useful ("unwrap failed in option.rs") so I don't think this makes any difference.
* However (the big win): When you see a (raw, unsymbolicated) stack trace in a log file or a bug report from a customer you can now symbolicate it to get a full stack trace. If you have a core dump, you can just fire up a debugger and you see everything!